Use lapack func instead of `scipy.linalg.cholesky`

Fyrebright opened the pull request:
* Now skips 2D checks in perform
* Updated the default arguments for `check_finite` to false to match documentation
* Add benchmark test case

<!-- !! Thank your for opening a PR !! -->

<!--- Provide a self-contained summary of your changes in the Title above -->
<!--- This is what will be shown in the automatic release notes: https://github.com/pymc-devs/pytensor/releases -->

## Description
<!--- Describe your changes in detail -->
Adds `_cholesky` method to `slinalg.Cholesky` to replace the `scipy.linalg.cholesky` wrapper. It is almost identical to [the corresponding scipy function](https://github.com/scipy/scipy/blob/63025cd1028cdb46de42400df21c60aa13f68743/scipy/linalg/_decomp_cholesky.py#L15) but it skips the 2d check and the batching wrapper.

Previous performance (with `check_finite=False`):
```
benchmark: 5.1.0 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)

--------------------------------------------------- benchmark: 1 tests ---------------------------------------------------
Name (time in us)                Min      Max    Mean  StdDev  Median     IQR   Outliers  OPS (Kops/s)  Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------------
test_cholesky_performance     5.6610  43.2910  6.8433  2.4983  5.9610  0.1810  1312;1574      146.1280    9487           1
--------------------------------------------------------------------------------------------------------------------------
```

After changes:
```
benchmark: 5.1.0 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)

--------------------------------------------------- benchmark: 1 tests --------------------------------------------------
Name (time in us)                Min      Max    Mean  StdDev  Median     IQR  Outliers  OPS (Kops/s)  Rounds  Iterations
-------------------------------------------------------------------------------------------------------------------------
test_cholesky_performance     5.1500  28.5640  5.5494  1.0101  5.4210  0.1100   262;628      180.2012   12753           1
-------------------------------------------------------------------------------------------------------------------------
```

## Related Issue
<!--- It is good practice to first open an issue explaining the bug / new feature that is addressed by this PR -->
<!--- Please type an `x` in one of the boxes below and provide the issue number after the # sign: -->
- [ ] Closes #
- [x] Related to #1468

## Checklist
<!--- Make sure you have completed the following steps before submitting your PR -->
<!--- Feel free to type an `x` in all the boxes below to let us know you have completed the steps: -->
- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)
<!--- You may find this guide helpful: https://mainmatter.com/blog/2021/05/26/keeping-a-clean-git-history/ -->

## Type of change
<!--- Select one of the categories below by typing an `x` in the box -->
- [ ] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [x] Other (please specify): performance
<!--- Additionally, if you are a maintainer or reviewer, please make sure that the appropriate labels are added to this PR -->


<!-- readthedocs-preview pytensor start -->
----
📚 Documentation preview 📚: https://pytensor--1487.org.readthedocs.build/en/1487/

<!-- readthedocs-preview pytensor end -->

ricardoV94 commented on pytensor/tensor/slinalg.py:
```diff
@@ -64,21 +65,52 @@ def make_node(self, x):
+    def _cholesky(
+        self, a, lower=False, overwrite_a=False, clean=True, check_finite=False
+    ):
+        a1 = np.asarray_chkfinite(a) if check_finite else np.asarray(a)
```
We can trust the input to always be numpy arrays in our case, which gets rid of `np.asarray` and the `_datacopied` below

ricardoV94 commented on pytensor/tensor/slinalg.py:
```diff
@@ -64,21 +65,52 @@ def make_node(self, x):
+
+        overwrite_a = overwrite_a or _datacopied(a1, a)
+        (potrf,) = scipy_linalg.get_lapack_funcs(("potrf",), (a1,))
+        c, info = potrf(a1, lower=lower, overwrite_a=overwrite_a, clean=clean)
+        if info > 0:
+            raise scipy_linalg.LinAlgError(
+                f"{info}-th leading minor of the array is not positive definite"
+            )
+        if info < 0:
+            raise ValueError(
+                f"LAPACK reported an illegal value in {-info}-th argument "
+                f'on entry to "POTRF".'
+            )
```
this depends `on_error` flag

Fyrebright replied:
I think the behavior was the same because of the try/catch, but it was clunky. It is simpler after refactoring out the helper.

ricardoV94 commented on pytensor/tensor/slinalg.py:
```diff
@@ -64,21 +65,52 @@ def make_node(self, x):
             # If we have a `C_CONTIGUOUS` array we transpose to benefit from it
             if self.overwrite_a and x.flags["C_CONTIGUOUS"]:
-                out[0] = scipy_linalg.cholesky(
+                out[0] = self._cholesky(
```
We can do this if/else logic inside our helper

ricardoV94 replied:
Actually you don't need a helper, just put the whole code in the perfom method

ricardoV94 reviewed (commented):
Nice start, left some commens.

Fyrebright commented:
Thank you for the feedback. I moved it all to `perform` and removed the `asarray`,`_datacopied` checks. I also moved the empty input case to the beginning and added test coverage for it.

jessegrabowski commented on pytensor/tensor/slinalg.py:
```diff
@@ -67,29 +67,52 @@ def make_node(self, x):
-            else:
+        # Quick return for square empty array
+        if x.size == 0:
+            eye = np.eye(1, dtype=x.dtype)
+            (potrf,) = scipy_linalg.get_lapack_funcs(("potrf",), (eye,))
+            c, _ = potrf(eye, lower=False, overwrite_a=False, clean=True)
+            out[0] = np.empty_like(x, dtype=c.dtype)
```
```suggestion
            (potrf,) = scipy_linalg.get_lapack_funcs(("potrf",), (x,))
            out[0] = np.empty_like(x, dtype=potrf.dtype)
```

Since you just want the dtype, you can check `potrf.dtype`, you don't need to actually call the routine. It's also fine to pass in the empty `x` in this case, since it's only used for its dtype attribute (the data or lack thereof doesn't matter)

jessegrabowski commented on pytensor/tensor/slinalg.py:
```diff
@@ -67,29 +67,52 @@ def make_node(self, x):
+            out[0] = np.empty_like(x, dtype=c.dtype)
+            return
+
+        x1 = np.asarray_chkfinite(x) if self.check_finite else x
```
I don't like the introduction of a new variable. We know for sure `x` is a numpy array inside this function, so the only relevant code in `np.asarray_chkfinite(x)` is the `if not np.isfinite(x).all(): ` part. We should just directly do that

Fyrebright replied:
Do you think that the `on_error` flag should affect whether or not a `ValueError` is raised? Or should it just be:
```suggestion
        if self.check_finite and not np.isfinite(x).all():
            raise ValueError("array must not contain infs or NaNs")
```

jessegrabowski replied:
Yes, `on_error` should effect whether an error raises. If `on_error='nan'` there should not be an error raised, it should return a matrix of the correct shape with all `nan`

jessegrabowski commented on pytensor/tensor/slinalg.py:
```diff
@@ -67,29 +67,52 @@ def make_node(self, x):
+        # Quick return for square empty array
+        if x.size == 0:
+            eye = np.eye(1, dtype=x.dtype)
+            (potrf,) = scipy_linalg.get_lapack_funcs(("potrf",), (eye,))
```
Get `potrf` once before this size check and re-use it in all branches

jessegrabowski commented on pytensor/tensor/slinalg.py:
```diff
@@ -67,29 +67,52 @@ def make_node(self, x):
+
+        # Scipy cholesky only makes use of overwrite_a when it is F_CONTIGUOUS
+        # If we have a `C_CONTIGUOUS` array we transpose to benefit from it
+        if self.overwrite_a and x.flags["C_CONTIGUOUS"]:
```
save a variable `c_contiguous_input` and use that for checks later (instead of maintaining two arrays)

jessegrabowski reviewed (changes requested):
Looking really great! I see a few more places we can improve the performance (avoiding copy + cleaning up the code) then I think it'll be ready to merge

jessegrabowski commented:
You have a test failure:

```
2025-06-21T21:14:33.0385589Z =================================== FAILURES ===================================
2025-06-21T21:14:33.0386251Z ______________________ test_cholesky_raises_on_nan_input _______________________
2025-06-21T21:14:33.0386688Z 
2025-06-21T21:14:33.0386858Z     def test_cholesky_raises_on_nan_input():
2025-06-21T21:14:33.0387381Z         test_value = rng.random(size=(3, 3)).astype(floatX)
2025-06-21T21:14:33.0387886Z         test_value[0, 0] = np.nan
2025-06-21T21:14:33.0388256Z     
2025-06-21T21:14:33.0388568Z         x = pt.tensor(dtype=floatX, shape=(3, 3))
2025-06-21T21:14:33.0389299Z         x = x.T.dot(x)
2025-06-21T21:14:33.0389718Z         g = pt.linalg.cholesky(x)
2025-06-21T21:14:33.0390184Z         f = pytensor.function([x], g, mode="NUMBA")
2025-06-21T21:14:33.0390622Z     
2025-06-21T21:14:33.0391099Z >       with pytest.raises(np.linalg.LinAlgError, match=r"Non-numeric values"):
2025-06-21T21:14:33.0391737Z              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2025-06-21T21:14:33.0392394Z E       Failed: DID NOT RAISE <class 'numpy.linalg.LinAlgError'>
2025-06-21T21:14:33.0392804Z 
2025-06-21T21:14:33.0392979Z tests/link/numba/test_slinalg.py:471: Failed
```

Could be related to the change in the default `check_finite` flag? You might need to adjust this test to correctly set the flag if so.

ricardoV94 commented on pytensor/tensor/slinalg.py:
```diff
@@ -67,29 +67,55 @@ def make_node(self, x):
+
+        if self.check_finite and not np.isfinite(x).all():
+            if self.on_error == "nan":
+                out[0] = np.full(x.shape, np.nan, dtype=node.outputs[0].type.dtype)
```
```suggestion
                out[0] = np.full(x.shape, np.nan, dtype=potrf.dtype)
```

ricardoV94 reviewed (commented):
Looks good. small comment

Fyrebright commented:
> Could be related to the change in the default `check_finite` flag? You might need to adjust this test to correctly set the flag if so.

It is fixed now. I had been skipping the `link/*` tests until now. They are all passing on my machine except for:
```
FAILED tests/scan/test_basic.py::test_cython_performance - ModuleNotFoundError: No module named 'pytensor.scan.scan_perform'
FAILED tests/scan/test_basic.py::test_output_storage_reuse[cvm] - ModuleNotFoundError: No module named 'pytensor.scan.scan_perform'
```

jessegrabowski reviewed (approved):
Looks great, thanks for this!
